proposals/p6331.md

```diff
@@ -0,0 +1,109 @@
+
+However, this mechanism falls short when a developer only needs C++ built-in
+types (for example, `long`, `long long`, `unsigned long`), as these are not
+defined in any header file.
+
+Furthermore, some C++ types are **nominally distinct**. For example, on many
+64-bit systems, `long` and `long long` share the same 64-bit representation, but
+C++ treats them as distinct types - a critical feature for function overloading.
+If Carbon were to map both types to its canonical `i64`, this semantic
+distinction would be lost, breaking interoperability.
+
+To address this, Carbon must provide its own distinct mirror types, such as
+`Cpp.long` and `Cpp.long_long`. The language currently lacks a clear and direct
+way to "activate" or import only these compiler-provided, built-in C++ entities.
+While `import Cpp inline "";` can achieve this, it is an unintuitive and
+indirect solution.
+
+## Background
+
+Currently, C++ built-in types can be brought into the `Cpp` namespace using
+`import Cpp inline "";`. This works because the `inline` import implicitly makes
+the built-in types available. However, this is a side effect of the `inline`
+import's behavior, not an obvious or intentional feature.
+
+## Proposal
+
+We propose a new import directive:
+
+```carbon
+import Cpp;
+```
+
+This file-level statement instructs the Carbon compiler to populate the `Cpp`
+namespace with all C++ built-in entities required for interoperability. This
+acts as a compiler-level activation and does not involve reading any files.
+
+## Details
+
+### Activated Entities
+
+The set of entities activated by `import Cpp;` is determined by the compiler's
+interoperability model. This will include C++ primitive types, including those
+necessary to preserve nominal distinction (for example, `Cpp.long`,
+`Cpp.unsigned_long`, `Cpp.long_long`).
+
+### Interaction with Other Imports
+
+The `import Cpp;` directive is **not** mutually exclusive with
+`import Cpp library "..."` or `import Cpp inline "..."`.
+
+Both `import Cpp library "..."` and `import Cpp inline "..."` will implicitly
+trigger the effects of `import Cpp;`. Since any C++ code is likely to use
+built-in types, making them available by default enhances the user experience.
+
+## Rationale
+
+**Explicitness over Magic:** This directive provides a clear, non-magical
+opt-in. It prevents the `Cpp` namespace from being automatically available to
+users who have not explicitly declared their intent to use C++ interoperability,
+aligning with Carbon's design principles.
+
+## Alternatives Considered
+
+1.  **Use `import Cpp inline "";`:** Rejected. This approach is not obvious and
+    relies on a side effect of the `inline` import implementation, rather than
+    being a clear, intentional feature.
+2.  **Implicit Activation:** Rejected. Automatically making `Cpp.long` available
+    in all files would be unexpected and would violate Carbon's design
+    principles.
```

Comment:
I wonder if there has been much discussion about this option?

These seem awfully like what we would want in a "prelude" like construct for C++. We already have a prelude that brings in names in `Core`, and the `Cpp` package name is going to be similarly special (or more special) to `Core` due to it enabling the C++ importing. For example, it isn't reasonable to define a Carbon file with the package name `Cpp`.

So maybe that argues that this should be part of Carbon's prelude, and that prelude should introduce names *both* in `Core` and `Cpp`?

(If this is controversial or we want a longer discussion, maybe this should be factored into a leads issue.)